graphql-server/src/schemas/custom.ts

```diff
@@ -487,5 +553,163 @@ function getAddlOtCodesMatchingCtOrCtdVals({
+
+    // Ensure isCmoSample is set in sample's 'additionalProperties' if not already present.
+    // This ensures that cmo samples get sent to the label generator after validation as
+    // some of the older SMILE samples do not have this additionalProperty set
+    if (sampleManifest["additionalProperties"]["isCmoSample"] == null) {
+      const requestId = sampleManifest["additionalProperties"]["igoRequestId"];
+      let req = ogm.model("Request");
+      const rd = await req.find({
+        where: { igoRequestId: requestId },
+      });
+      sampleManifest["additionalProperties"]["isCmoSample"] =
+        rd[0]["isCmoRequest"].toString();
+    }
+
+    publishNatsMessage(
+      props.pub_validate_sample_update,
+      JSON.stringify(sampleManifest)
+    );
+
+    await ogm.model("Sample").update({
+      where: { smileSampleId: sampleManifest.smileSampleId },
+      update: { revisable: false },
+    });
+
+    resolve(null);
+  });
+}
+
+async function updateTempo(newDashboardSample: DashboardSampleInput) {
+  return new Promise((resolve) => {
+    const dataForTempoBillingUpdate = {
+      primaryId: newDashboardSample.primaryId,
+      billed: newDashboardSample.billed,
+      billedBy: newDashboardSample.billedBy,
+      costCenter: newDashboardSample.costCenter,
+      accessLevel: newDashboardSample.accessLevel,
+      custodianInformation: newDashboardSample.custodianInformation,
+    };
+
+    publishNatsMessage(
+      props.pub_tempo_sample_billing,
+      JSON.stringify(dataForTempoBillingUpdate)
+    );
+
+    resolve(null);
+  });
+}
+
+const editableSampleMetadataFields = new Set([
+  "cmoPatientId",
+  "investigatorSampleId",
+  "sampleType",
+  "preservation",
+  "tumorOrNormal",
+  "sampleClass",
+  "oncotreeCode",
+  "collectionYear",
+  "sampleOrigin",
+  "tissueLocation",
+]);
+
+const editableTempoFields = new Set([
+  "billed",
+  "costCenter",
+  "billedBy",
+  "custodianInformation",
+  "accessLevel",
+]);
+
+async function updateAllSamplesConcurrently(
+  newDashboardSamples: DashboardSampleInput[],
+  ogm: OGM
+) {
+  const allPromises = newDashboardSamples.map(async (dashboardSample) => {
+    try {
+      const metadataChanged = dashboardSample.changedFieldNames.some((field) =>
+        editableSampleMetadataFields.has(field)
+      );
+      const tempoChanged = dashboardSample.changedFieldNames.some((field) =>
+        editableTempoFields.has(field)
+      );
+
+      const promises = [];
+      if (metadataChanged) {
+        promises.push(updateSampleMetadata(dashboardSample, ogm));
+      }
+      if (tempoChanged) {
+        promises.push(updateTempo(dashboardSample));
+      }
+
+      return Promise.all(promises);
+    } catch (error) {
+      console.error(
+        `Failed to update sample with primaryId ${dashboardSample.primaryId}. Error:`,
+        error
+      );
+      throw error; // ensure Promise.allSettled captures the error
+    }
+  });
+
+  await Promise.allSettled(allPromises);
+}
```

Comment:
Process all the new sample updates concurrently (_not_ in parallel) to boost performance. If the sample being updated contains a SampleMetadata update, then send a NATS message to the "igo sample update" topic. Same thing with the "tempo sample billing" topic.